### PR TITLE
Add CI

### PR DESCRIPTION
**Pull request by constantinpape**

### CODE REVIEW

**Comment by constantinpape** [comment]
Started to set up the CI now. Still have to add skimage to the dependencies.
Also, when I  use `make test` to run the `read_all` test, it triggers the data generation, because
`data` is listed as target here:
https://github.com/constantinpape/zarr_implementations/blob/master/Makefile#L23-L24
@clbarnes I guess you have intended this when adding the make file to test the data generation.
But I am not sure if this will even work, because the data is already there?

**Comment by clbarnes** [comment]
Yeah, I wanted to express the dependency of the tests on the data. It would be nice to actually use make's file-checking mechanisms but I don't know if it works with directories, or generally with multiple files. It's quite old-fashioned.

I think that the CI should try to generate the data to at least once anyway, as a separate test - then do a `git diff --exit-code --name-only data`, to prove that the data in the repo is exactly what is created by the scripts in the repo. Scripts and data which are out of sync should be a failure.

If we go that route, then we'll have installed all the dependencies and compiled everything in order to run that test, in which case regenerating the data for the read tests should be a negligible amount of time lost.

**Comment by constantinpape** [comment]
> I think that the CI should try to generate the data to at least once anyway, as a separate test - then do a `git diff --exit-code --name-only data`, to prove that the data in the repo is exactly what is created by the scripts in the repo. Scripts and data which are out of sync should be a failure.

I agree, that makes sense. But we need to generate the data in a different folder for this and this is not supported by the current data generation scripts, because all paths are hard-coded. Should be easy to fix this though.

> If we go that route, then we'll have installed all the dependencies and compiled everything in order to run that test, in which case regenerating the data for the read tests should be a negligible amount of time lost. But we could always just remove the `data` recipe from the dependencies of the `test` recipe.

Yes, I am also not worried about runtime here.

**Comment by clbarnes** [comment]
> But we need to generate the data in a different folder for this

I'm not sure we do - doing it in place makes it easy for git to do the diff for us, and as it's just done on CI any changes get trashed anyway.

**Comment by constantinpape** [comment]
> I'm not sure we do - doing it in place makes it easy for git to do the diff for us, and as it's just done on CI any changes get trashed anyway.

Ok, that's a good point.

However some of the scripts fail if they are called with existing data, e.g.
https://github.com/constantinpape/zarr_implementations/blob/master/generate_data/generate_z5py.py#L25

I will fix it on the python side now and see if it works in java.

**Comment by clbarnes** [comment]
Ah, right. Could always `rm -rf` the data in there first as part of the `data` recipe: that way we can be sure that the generated data and nothing but the generated data is in that directory. Alternatively, the scripts could open their containers in write-truncate mode (`mode="w"`) to clear previous writes. That's probably a good idea anyway to make sure defunct datasets don't hang around if support for particular compressors change in the future.

File: generate_data/n5-java/generate_data.sh [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# clean the n5 data. ideally, this would be done in n5-java, but I don't know if
+# it supports truncate mode
+rm -r data/n5-java.n5
```
Comment (clbarnes) [inline generate_data/n5-java/generate_data.sh:3]: Could this go under the `cd` line, using the relative path `../../data/n5-java.n5`? Then it stays agnostic as to where it's called from.

Comment (constantinpape) [reply generate_data/n5-java/generate_data.sh:3]: Yes, makes sense.

**Comment by constantinpape** [comment]
I updated the truncate in n5 as you suggested @clbarnes.
Note that generating the reference image currently fails due to https://github.com/python-pillow/Pillow/issues/4518.
This is fixed in the latest pillow release, however this is not on conda-forge yet, so this PR is pending https://github.com/conda-forge/pillow-feedstock/pull/78.

**Comment by constantinpape** [comment]
Ok, the environment can be set up now and the tests are running; a couple are still failing though, may also be related to #7. I will see if I can reproduce and fix it locally later.

**Comment by hugovk** [comment]
https://github.com/conda-forge/pillow-feedstock/pull/78 is now merged.

**Comment by constantinpape** [comment]
> [conda-forge/pillow-feedstock#78](https://github.com/conda-forge/pillow-feedstock/pull/78) is now merged.

Yes, works here again. Test failures are due some other issues.

@clbarnes 
This is some z5py related things; I don't have time to fix this now, will try to look into it next week.
If it's helpful for you I can merge already, otherwise I will wait till I fixed these issues.

**Comment by clbarnes** [comment]
No hurry at all! Just thought it would be useful for this repo going forward.

**Comment by constantinpape** [comment]
> No hurry at all! Just thought it would be useful for this repo going forward.

Yes, definitely!
I will see if I have time to look into it next week, otherwise I will just merge.

**Comment by constantinpape** [comment]
Superseded by #14

**Pull request closed without merging**